File: By-Swift/By-Swift/Programmers/[2021-DevMatching]/다단계칫솔판매.swift
import Foundation

func solution(_ enroll:[String], _ referral:[String], _ seller:[String], _ amount:[Int]) -> [Int] {
    var relation = [String: String]()
    var score = [String: Int]()
    
    for i in 0..<enroll.count {
        relation[enroll[i]] = referral[i]
        score[enroll[i]] = 0
    }
    
    for i in 0..<seller.count {
        var cur = seller[i]
        var money = amount[i]*100
        var tax = Int(Double(money)*0.1)
        while cur != "-" {
            if tax >= 1 {
                score[cur]! += money-tax
            } else {
                score[cur]! += money
                break
            }
            
            cur = relation[cur]!
            money = tax
            tax = Int(Double(money)*0.1)
        }
    }
    
    var answer = [Int]()
    for i in enroll {
        answer.append(score[i]!)
    }
    return answer
}

File: By-Python/programmers/[2021-DevMatching]/다단계칫솔판매.py
def solution(enroll, referral, seller, amount):
    relation = dict()
    score = {"center" : 0}
    
    for i in range(len(enroll)):
        if referral[i] == "-":
            referral[i] = "center"
        
        relation[enroll[i]] = referral[i]
        score[enroll[i]] = 0
    

    # s: 판매자 , money: 얻은 수익(수수료 떼야 될 금액)
    def dfs(s, money):
        tax = int(money*0.1)
        if tax >= 1:
                score[s] += (money-tax)
        else: 
            score[s] += money
            return

        if s == "center": 
            return
        else: 
            return dfs(relation[s], tax)

        
    for i in range(len(seller)):
        dfs(seller[i], amount[i]*100)
            
    del score["center"]
    return list(score.values())